File: javascript/js-client/src/protocol/crypto.js
import { Sha256 } from "./sha256.js";

const _P = 2n ** 255n - 19n;
const _A24 = 121665n;
const _CHACHA_CONSTS = [0x61707865, 0x3320646e, 0x79622d32, 0x6b206574];

function _rotl32(x, n) {
    return ((x << n) | (x >>> (32 - n))) >>> 0;
}

function _quarterRound(s, a, b, c, d) {
    s[a] = (s[a] + s[b]) >>> 0;
    s[d] = _rotl32(s[d] ^ s[a], 16);
    s[c] = (s[c] + s[d]) >>> 0;
    s[b] = _rotl32(s[b] ^ s[c], 12);
    s[a] = (s[a] + s[b]) >>> 0;
    s[d] = _rotl32(s[d] ^ s[a], 8);
    s[c] = (s[c] + s[d]) >>> 0;
    s[b] = _rotl32(s[b] ^ s[c], 7);
}

function _chachaBlock(state) {
    const out = new Int32Array(state);
    for (let i = 0; i < 10; i++) {
        _quarterRound(out, 0, 4, 8, 12);
        _quarterRound(out, 1, 5, 9, 13);
        _quarterRound(out, 2, 6, 10, 14);
        _quarterRound(out, 3, 7, 11, 15);
        _quarterRound(out, 0, 5, 10, 15);
        _quarterRound(out, 1, 6, 11, 12);
        _quarterRound(out, 2, 7, 8, 13);
        _quarterRound(out, 3, 4, 9, 14);
    }
    const ks = new Uint8Array(64);
    const dv = new DataView(ks.buffer);
    for (let i = 0; i < 16; i++) {
        dv.setUint32(i * 4, (out[i] + state[i]) >>> 0, true);
    }
    return ks;
}

export function chachaCrypt(data, key, packetIndex, decrypt) {
    const state = new Int32Array(16);
    for (let i = 0; i < 4; i++) state[i] = _CHACHA_CONSTS[i];
    const kd = new DataView(key.buffer, key.byteOffset, key.byteLength);
    for (let i = 0; i < 8; i++) state[4 + i] = kd.getInt32(i * 4, true);
    state[13] = 0;
    state[14] = packetIndex | 0;
    state[15] = decrypt ? -2147483648 : 0;
    const out = new Uint8Array(data.length);
    for (let i = 0; i < data.length; i += 64) {
        state[12] = (i / 64) | 0;
        const ks = _chachaBlock(state);
        for (let j = 0; j < 64 && i + j < data.length; j++) {
            out[i + j] = data[i + j] ^ ks[j];
        }
    }
    return out;
}

function _u64le(n) {
    const b = new Uint8Array(8);
    const dv = new DataView(b.buffer);
    dv.setUint32(0, n >>> 0, true);
    dv.setUint32(4, Math.floor(n / 2 ** 32) >>> 0, true);
    return b;
}

export class ArrasCipher {
    constructor(sharedKey) {
        this.key = sharedKey instanceof Uint8Array ? sharedKey : new Uint8Array(sharedKey);
        this.sent = 0;
        this.received = 0;
    }

    encrypt(packet) {
        const idx = this.sent++;
        const cipher = chachaCrypt(packet, this.key, idx, false);
        const hasher = new Sha256();
        hasher.update(cipher);
        hasher.update(this.key);
        hasher.update(_u64le(idx));
        const mac = hasher.digest().subarray(0, 6);
        const result = new Uint8Array(cipher.length + 6);
        result.set(cipher);
        result.set(mac, cipher.length);
        return result;
    }

    decrypt(packet) {
        const idx = this.received++;
        return chachaCrypt(packet, this.key, idx, true);
    }
}

function _modpow(base, exp, mod) {
    let result = 1n;
    base %= mod;
    while (exp > 0n) {
        if (exp & 1n) result = (result * base) % mod;
        base = (base * base) % mod;
        exp >>= 1n;
    }
    return result;
}

export function x25519(k, u) {
    const kb = new Uint8Array(32);
    kb.set((k instanceof Uint8Array ? k : new Uint8Array(k)).slice(0, 32));
    kb[0] &= 248;
    kb[31] &= 127;
    kb[31] |= 64;
    const uv = u instanceof Uint8Array ? u : new Uint8Array(u);
    let hex = "";
    for (let i = 31; i >= 0; i--) hex += uv[i].toString(16).padStart(2, "0");
    const x1 = BigInt("0x" + hex);
    let x2 = 1n, z2 = 0n, x3 = x1, z3 = 1n, swap = 0n;
    for (let t = 254; t >= 0; t--) {
        const kt = BigInt((kb[t >> 3] >> (t & 7)) & 1);
        swap ^= kt;
        if (swap) { [x2, x3] = [x3, x2]; [z2, z3] = [z3, z2]; }
        swap = kt;
        const a = (x2 + z2) % _P;
        const aa = (a * a) % _P;
        const b = (x2 - z2) % _P;
        const bb = (b * b) % _P;
        const e = (aa - bb) % _P;
        const c = (x3 + z3) % _P;
        const d = (x3 - z3) % _P;
        const da = (d * a) % _P;
        const cb = (c * b) % _P;
        const sum = (da + cb) % _P;
        const diff = ((da - cb) % _P + _P) % _P;
        x3 = (sum * sum) % _P;
        z3 = (diff * diff * x1) % _P;
        x2 = (aa * bb) % _P;
        z2 = (e * ((aa + _A24 * e) % _P)) % _P;
    }
    if (swap) { [x2, x3] = [x3, x2]; [z2, z3] = [z3, z2]; }
    const res = (x2 * _modpow(z2, _P - 2n, _P)) % _P;
    const hexStr = res.toString(16).padStart(64, "0");
    const out = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
        out[i] = parseInt(hexStr.slice((31 - i) * 2, (31 - i) * 2 + 2), 16);
    }
    return out;
}

export function x25519Base(k) {
    const base = new Uint8Array(32);
    base[0] = 9;
    return x25519(k, base);
}

File: python/arras/protocol/crypto.py
import hashlib

P = 2 ** 255 - 19
A24 = 121665

CHACHA_CONST = (0x61707865, 0x3320646e, 0x79622d32, 0x6b206574)


def x25519(k: bytes, u: bytes) -> bytes:
    k = bytearray(k[:32].ljust(32, b"\x00"))
    k[0] &= 248
    k[31] &= 127
    k[31] |= 64
    x1 = int.from_bytes(u[:32], "little")
    x2, z2 = 1, 0
    x3, z3 = x1, 1
    swap = 0
    for t in range(254, -1, -1):
        kt = (k[t // 8] >> (t % 8)) & 1
        swap ^= kt
        if swap:
            x2, x3 = x3, x2
            z2, z3 = z3, z2
        swap = kt
        a = (x2 + z2) % P
        aa = a * a % P
        b = (x2 - z2) % P
        bb = b * b % P
        e = (aa - bb) % P
        c = (x3 + z3) % P
        d = (x3 - z3) % P
        da = d * a % P
        cb = c * b % P
        x3 = (da + cb) % P
        x3 = x3 * x3 % P
        z3 = (da - cb) % P
        z3 = z3 * z3 % P
        z3 = z3 * x1 % P
        x2 = aa * bb % P
        z2 = e * (aa + A24 * e) % P
    if swap:
        x2, x3 = x3, x2
        z2, z3 = z3, z2
    return (x2 * pow(z2, P - 2, P) % P).to_bytes(32, "little")

def x25519_base(k: bytes) -> bytes:
    return x25519(k, (9).to_bytes(32, "little"))

def _rotl32(x: int, n: int) -> int:
    return ((x << n) | (x >> (32 - n))) & 0xFFFFFFFF

def _quarter_round(s, a, b, c, d):
    s[a] = (s[a] + s[b]) & 0xFFFFFFFF
    s[d] = _rotl32(s[d] ^ s[a], 16)
    s[c] = (s[c] + s[d]) & 0xFFFFFFFF
    s[b] = _rotl32(s[b] ^ s[c], 12)
    s[a] = (s[a] + s[b]) & 0xFFFFFFFF
    s[d] = _rotl32(s[d] ^ s[a], 8)
    s[c] = (s[c] + s[d]) & 0xFFFFFFFF
    s[b] = _rotl32(s[b] ^ s[c], 7)

def _chacha_block(state):
    out = list(state)
    for _ in range(10):
        _quarter_round(out, 0, 4, 8, 12)
        _quarter_round(out, 1, 5, 9, 13)
        _quarter_round(out, 2, 6, 10, 14)
        _quarter_round(out, 3, 7, 11, 15)
        _quarter_round(out, 0, 5, 10, 15)
        _quarter_round(out, 1, 6, 11, 12)
        _quarter_round(out, 2, 7, 8, 13)
        _quarter_round(out, 3, 4, 9, 14)
    return [(out[i] + state[i]) & 0xFFFFFFFF for i in range(16)]

def _state(key: bytes, packet_index: int, decrypt: bool) -> list:
    words = [int.from_bytes(key[i:i + 4], "little") for i in range(0, 32, 4)]
    return list(CHACHA_CONST) + words + [
        0,
        0,
        packet_index & 0xFFFFFFFF,
        (packet_index >> 32) & 0xFFFFFFFF if not decrypt else 0x80000000,
    ]

def _crypt(data: bytes, key: bytes, packet_index: int, decrypt: bool) -> bytes:
    out = bytearray(len(data))
    state = _state(key, packet_index, decrypt)
    for i in range(0, len(data), 64):
        state[12] = i // 64
        ks = _chacha_block(state)
        for j in range(64):
            if i + j < len(data):
                out[i + j] = data[i + j] ^ ((ks[j // 4] >> (8 * (j % 4))) & 0xFF)
    return bytes(out)

class ArrasCipher:
    def __init__(self, shared_key: bytes):
        self.key = shared_key
        self.sent = 0
        self.received = 0

    def encrypt(self, packet: bytes) -> bytes:
        idx = self.sent
        self.sent += 1
        cipher = _crypt(packet, self.key, idx, decrypt=False)
        mac = hashlib.sha256(cipher + self.key + idx.to_bytes(8, "little")).digest()
        return cipher + mac[:6]

    def decrypt(self, packet: bytes) -> bytes:
        idx = self.received
        self.received += 1
        return _crypt(packet, self.key, idx, decrypt=True)

if __name__ == "__main__":
    alice_priv = bytes.fromhex("77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a")
    alice_pub = bytes.fromhex("8520f0098930a754748b7ddcb43ef75a0dbf3a0d26381af4eba4a98eaa9b4e6a")
    bob_priv = bytes.fromhex("5dab087e624a8a4b79e17f8b83800ee66f3bb1292618b6fd1c2f8b27ff88e0eb")
    bob_pub = bytes.fromhex("de9edb7d7b7dc1b4d35b61c2ece435373f8343c85b78674dadfc7e146f882b4f")
    shared = bytes.fromhex("4a5d9d5ba4ce2de1728e3bf480350f25e07e21c947d19e3376f09b3c1e161742")
    assert x25519(alice_priv, bob_pub) == shared, "alice x25519 mismatch"
    assert x25519(bob_priv, alice_pub) == shared, "bob x25519 mismatch"
    assert x25519_base(alice_priv) == alice_pub, "alice base mismatch"
    assert x25519_base(bob_priv) == bob_pub, "bob base mismatch"
    print("x25519 RFC 7748 vectors: OK")

    key = bytes(range(32))
    c = ArrasCipher(key)
    pkt = bytes([0x6B, 0xC0, 0xC0, 0xC0])
    enc = c.encrypt(pkt)
    assert len(enc) == len(pkt) + 6
    c2 = ArrasCipher(key)
    dec = c2.decrypt(enc)
    assert dec == pkt, f"roundtrip mismatch: {dec.hex()}"
    print("encrypt/decrypt roundtrip: OK")
